From ff3e0ad852bd443bda8cfaf46ed739d7793f9219 Mon Sep 17 00:00:00 2001
From: Liu Ying <victor.liu@nxp.com>
Date: Thu, 30 Aug 2018 16:13:49 +0800
Subject: [PATCH 4537/5242] MLK-19413-26 gpu: imx: dpu: framegen: Support two
 helpers for secondary syncup status

commit  68b2c08cc2cae28dc823ad6177e38ab04a77a0ab from
https://source.codeaurora.org/external/imx/linux-imx.git

This patch adds framegen_secondary_is_syncup() and
framegen_wait_for_secondary_syncup() helpers support so that
the callers may know a framegen's syncup status for the
secondary input.

Signed-off-by: Liu Ying <victor.liu@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/gpu/imx/dpu/dpu-framegen.c |   31 +++++++++++++++++++++++++++++++
 include/video/dpu.h                |    2 ++
 2 files changed, 33 insertions(+)

diff --git a/drivers/gpu/imx/dpu/dpu-framegen.c b/drivers/gpu/imx/dpu/dpu-framegen.c
index 06b5b55..93a1b38 100644
--- a/drivers/gpu/imx/dpu/dpu-framegen.c
+++ b/drivers/gpu/imx/dpu/dpu-framegen.c
@@ -78,6 +78,7 @@
 #define FRAMEINDEX_MASK		0xFFFFC000
 #define FRAMEINDEX_SHIFT	14
 #define FGCHSTAT		0x78
+#define SECSYNCSTAT		BIT(24)
 #define FGCHSTATCLR		0x7C
 #define FGSKEWMON		0x80
 #define FGSFIFOMIN		0x84
@@ -581,6 +582,36 @@ void framegen_wait_for_frame_counter_moving(struct dpu_framegen *fg)
 }
 EXPORT_SYMBOL_GPL(framegen_wait_for_frame_counter_moving);
 
+bool framegen_secondary_is_syncup(struct dpu_framegen *fg)
+{
+	u32 val;
+
+	mutex_lock(&fg->mutex);
+	val = dpu_fg_read(fg, FGCHSTAT);
+	mutex_unlock(&fg->mutex);
+
+	return val & SECSYNCSTAT;
+}
+EXPORT_SYMBOL_GPL(framegen_secondary_is_syncup);
+
+void framegen_wait_for_secondary_syncup(struct dpu_framegen *fg)
+{
+	unsigned long timeout = jiffies + msecs_to_jiffies(50);
+	bool syncup;
+
+	do {
+		syncup = framegen_secondary_is_syncup(fg);
+	} while (!syncup && time_before(jiffies, timeout));
+
+	if (syncup)
+		dev_dbg(fg->dpu->dev, "FrameGen%d secondary syncup\n", fg->id);
+	else
+		dev_err(fg->dpu->dev,
+			"failed to wait for FrameGen%d secondary syncup\n",
+			fg->id);
+}
+EXPORT_SYMBOL_GPL(framegen_wait_for_secondary_syncup);
+
 void framegen_enable_clock(struct dpu_framegen *fg)
 {
 	if (!fg->use_bypass_clk)
diff --git a/include/video/dpu.h b/include/video/dpu.h
index 96202855..f679d88 100644
--- a/include/video/dpu.h
+++ b/include/video/dpu.h
@@ -621,6 +621,8 @@ void fetchwarp_rgb_constantcolor(struct dpu_fetchunit *fu,
 void framegen_read_timestamp(struct dpu_framegen *fg,
 			     u32 *frame_index, u32 *line_index);
 void framegen_wait_for_frame_counter_moving(struct dpu_framegen *fg);
+bool framegen_secondary_is_syncup(struct dpu_framegen *fg);
+void framegen_wait_for_secondary_syncup(struct dpu_framegen *fg);
 void framegen_enable_clock(struct dpu_framegen *fg);
 void framegen_disable_clock(struct dpu_framegen *fg);
 bool framegen_is_master(struct dpu_framegen *fg);
-- 
1.7.9.5

